Issue #802: MyRocks: Statement rollback doesnt work correctly for nes…

spetrunia opened the pull request:
…ted statements

Variant #1: When the statement fails, we should roll back to the latest
savepoint taken at the top level.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -2449,6 +2453,10 @@ class Rdb_transaction_impl : public Rdb_transaction {
 
+  // Number of savepoints to roll back to get to the start of the
+  // "top-level" statement.
+  int m_n_savepoints;
```
It would probably make more sense to have these variables and collapse_savepoint() call in Rdb_transaction instead?

spetrunia committed:
Issue #802: MyRocks: Statement rollback doesnt work correctly for nested statements

Variant#2: Use a different approach to taking RocksDB SavePoints.
- Take one at transaction start (so we can rollback to it)
- Take one when top-level statement completes (there is a place to do
  that. Detecting when a top-level statement starts is much harder)

- Also, don't make a new Savepoint if the previous statement didn't
  make any writes. This way, a long transaction doing lots of SELECTs
  will not create a lot of savepoints.

- Also, don't call get_or_create_tx() from
  ha_rocksdb::external_lock(F_UNLCK). This call is made right after
  rocksdb_commit(). We end up creating a new RocksDB transaction just
  to do cleanups and immediately discard it.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -1854,6 +1854,17 @@ class Rdb_transaction {
 
+private:
+  /* Number of RockDB savepoints taken */
+  int m_n_savepoints;
```
m_n_savepoints and m_writes_at_last_savepoint need to be initialized at similar locations to m_write_count, no?

spetrunia replied:
Yes. Lost that initialization when re-working the patch. Will now add it back

spetrunia committed:
Issue #802: MyRocks: Statement rollback doesnt work correctly for nested statements

Variant #2: initialize m_n_savepoints and m_writes_at_last_savepoint
at transaction start.
